### ISSUE TITLE
Scripting API: Missing tileset "firstGID" property, and flip flag values don't match

### ISSUE THREAD

**Issue by rgrams** [mapeditor/tiled#3189]
I'm trying to sort out the conversion to and from tile IDs and GIDs, but the scripting API seems designed to prevent that.

__First GID__
Documentation, issues, and forum posts all mention how to use the tileset's "first GID". It's easy, it makes sense, you just add and subtract. However, in the scripting API, tilesets are missing the firstGID property, instead they have "nextTileID" - the tile index after the last one used—in other words, the _next_ tileset's firstGID...???

__Flag Bits__
It's the same with the flag bits, all the information you can find says that the flags for flipped/rotated tiles use the _highest_ four bits (for obvious reasons, so they don't intersect with local tile ID numbers), but the [flag variables](https://www.mapeditor.org/docs/scripting/classes/tile.html#flippedantidiagonally) that are part of the scripting API use the _lowest_ four bits.

I'm sure there must be reasons for this, but for me it's just adding hours of frustration for no reason. The scripting API seems to provide the _exact opposite_ of what I want (and what the docs say you need). Possibly the solution is just to improve the documentation. Quite a lot of information that you need to import and export to and from Tiled (and for scripting in general) is missing or just very hard to find.

**Comment by eishiya** [comment]
Everything you described only  applies to **file output** such as TMX and JSON, which use global IDs. These global IDs are only generated when the files are written, they are irrelevant and do not exist in the abstract runtime representations, which deal directly with Maps, Tilesets, and Tiles.

`firstgid` is needed to figure out which tileset a given tile is from when using global IDs. In scripting, you don't need this because you can get the tileset from the Tile with `tile.tileset`.

The flag bits aren't an inherent part of the tile, they're part of the map data. You can get the flags for a given cell in a layer with `layer.flagsAt(x, y)`, and you can check whether a given flag is set by ANDing the value with the [flip constants in the Tile class](https://www.mapeditor.org/docs/scripting/classes/tile.html#flippedantidiagonally).

**Comment by bjorn** [comment]
While everything @eishiya wrote is correct and I hope it helps you, @rgrams, I could imagine parts of the documentation or scripting API reference could be written in a way that makes this more clear.

Also, since global tile IDs are used by the built-in TMX, JSON and Lua file formats, I could imagine this system being useful for other custom formats as well, so potentially it makes sense to expose a helper API for converting between cells and global IDs (in the C++ code this is done by the `GidMapper`). I think a pure JavaScript implementation would likely be preferable though, also because then it can be customized when needed. It could be provided as an example somewhere.

**Comment by rgrams** [comment]
Thanks for taking the time to make that writeup. I did actually figure it out and get it working before I made this issue, the problem is how long it took to get there. What should have been clear and taken 20 minutes instead took several hours of searching and experimenting.

It doesn't really matter why the Tiled runtime doesn't use them or why scripts don't need them, because A) I _do_ need them. On both ends: To interpret data from Tiled in my game, and to write my own exporter that is similar to the built-in ones, and B) It's confusing that Tiled uses different values for the same things in different places. Formats that Tiled exports to out-of-the-box are just as much a part of Tiled as the scripting API.

@bjorn Are the files for the [scripting API docs](https://www.mapeditor.org/docs/scripting/index.html) up for public contribution somewhere? I found the files for the manuals in this repo, but not the API docs.

**Comment by bjorn** [comment]
> @bjorn Are the files for the [scripting API docs](https://www.mapeditor.org/docs/scripting/index.html) up for public contribution somewhere? I found the files for the manuals in this repo, but not the API docs.

Yes, the scripting API is based on [this TypeScript file](https://github.com/mapeditor/tiled/blob/master/docs/scripting-doc/index.d.ts) and any contributions are welcome!
